### PR TITLE
add sample app that allows you to easily compare maps provided by Esri and HERE in Amazon Location

### PR DESCRIPTION
**Pull request by inariku**
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

## Abstract
I have created a sample app that allows you to easily compare maps provided by Esri and HERE in Amazon Location. 
I had a customer who wanted to try out the Esri and HERE maps. The customer can see the images of the maps from the following link, but cannot try them out, so I created this app.
https://aws.amazon.com/jp/location/data-providers/esri/

## Demo site
I have a demo site, so please feel free to try it out.:arrow_right: [link](https://main.d2h14o0tvpo3e0.amplifyapp.com/)

### CODE REVIEW

**Comment by dreamorosi** [comment]
Hi, thank you for the contribution.

I see that you have added a link/submodule instead of the files for the sample
image


Could you please add the files instead? Ideally we'd like to have this repo not depending on external ones when it comes to code.

**Comment by inariku** [comment]
Sorry... I just added another file. Did I solve the trouble?

**Comment by dreamorosi** [comment]
Thank you, yes, now it seems right.

Please allow us some time to review the PR and give you feedback :)

**Comment by zachelliottwx** [comment]
Couple notes on Installation on a fresh Cloud9 instance
1: Yarn may not be installed by default. I had to install it.
2: Should include steps on how to do amplify init. New people may not know the options you need to select when doing the init command
3: Ran into an issue that vue cli was not installed. Even after attempting to install it, I still get errors and cannot get a successful installation.

Notes on the app: 
Is there anyway to sync the mouse/zoom between maps? I'm not an Amplify expert so i'm not sure how you could do that. Maybe @dreamorosi  has an idea. But having the two in sync would be awesome for comparison purposes and would make me feel better about releasing this into the wild.

**Comment by inariku** [comment]
@zachelliottwx Thank you. I have corrected the points you made. I now know how to sync the two maps. However, I ran into a problem where the new app works locally, but not when hosted.
Please give me some time to work on this problem.

**Comment by inariku** [comment]
@dreamorosi 
Sorry for the delay. The map sync feature now works well. Please review. You can try the new features at the link below.

https://main.dsv6j39h42g17.amplifyapp.com/

**Review by dreamorosi — CHANGES_REQUESTED** [review]
Additional notes:
- The image used in the readme is very big (almost 4MB) - this will affect the time it takes to `git clone` the whole repo. Please optimize it for web with something [like this tool](https://imagecompressor.com).
- Remove the `yarn.lock` file, I saw `yarn` was used in a previous version of the PR but since now we are using only `npm` the file is redundant

----

Finally, regarding the demo itself, this is how it's displayed on my screen:
Screenshot 2022-07-30 at 14 29 31

Is there any chance that we can make the maps bigger (i.e. remove the extra margins around) + make them the same height (if you see in my case the left is a bit smaller)?

File: maplibre-gl-js-vue-amplify-compare-maps/src/aws-exports.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
+/* eslint-disable */
+// WARNING: DO NOT EDIT. This file is automatically generated by AWS Amplify. It will be overwritten.
+
+const awsmobile = {
```
Comment (dreamorosi) [inline maplibre-gl-js-vue-amplify-compare-maps/src/aws-exports.js:4]: This file should never be committed to version control and definitely should not be part of this repo, please remove it asap.

Comment (inariku) [reply maplibre-gl-js-vue-amplify-compare-maps/src/aws-exports.js]: Done

Comment (inariku) [reply maplibre-gl-js-vue-amplify-compare-maps/src/aws-exports.js]: Done

File: maplibre-gl-js-vue-amplify-compare-maps/README.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
+# Amazon Location Map Comparison
+![](./image/README.png)
+
+This sample is used to compare maps provided by Amazon Location.
+## Execution environment
```
[the remark this answers is not in the document]
Comment (inariku) [reply maplibre-gl-js-vue-amplify-compare-maps/README.md:8]: Done

File: maplibre-gl-js-vue-amplify-compare-maps/README.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+* node v16.x
+* npm 8.x
+
+## Getting started
+1. Install Amplify CLI: `npm install -g @aws-amplify/cli`
+1. Install project dependencies: `npm install`
+1. Create a new Amplify project environment: `amplify init`
```
[the remark this answers is not in the document]
Comment (inariku) [reply maplibre-gl-js-vue-amplify-compare-maps/README.md:18]: Done

File: maplibre-gl-js-vue-amplify-compare-maps/README.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
+# Amazon Location Map Comparison
+![](./image/README.png)
+
+This sample is used to compare maps provided by Amazon Location.
+## Execution environment
+* node v16.x
+* npm 8.x
+
+## Getting started
+1. Install Amplify CLI: `npm install -g @aws-amplify/cli`
```
[the remark this answers is not in the document]
Comment (inariku) [reply maplibre-gl-js-vue-amplify-compare-maps/README.md:16]: Done

File: maplibre-gl-js-vue-amplify-compare-maps/README.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
+# Amazon Location Map Comparison
+![](./image/README.png)
+
+This sample is used to compare maps provided by Amazon Location.
+## Execution environment
+* node v16.x
+* npm 8.x
```
[the remark this answers is not in the document]
Comment (inariku) [reply maplibre-gl-js-vue-amplify-compare-maps/README.md:10]: Done

File: maplibre-gl-js-vue-amplify-compare-maps/README.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
+# Amazon Location Map Comparison
+![](./image/README.png)
+
+This sample is used to compare maps provided by Amazon Location.
```
[the remark this answers is not in the document]
Comment (inariku) [reply maplibre-gl-js-vue-amplify-compare-maps/README.md]: Done

File: maplibre-gl-js-vue-amplify-compare-maps/README.md [hunk 7]
```diff
@@ -0,0 +1,34 @@
+# Amazon Location Map Comparison
```
[the remark this answers is not in the document]
Comment (inariku) [reply maplibre-gl-js-vue-amplify-compare-maps/README.md]: Done

File: maplibre-gl-js-vue-amplify-compare-maps/README.md [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+## Cleaning up
+```
+amplify delete
+```
+
+## Security
+See [CONTRIBUTING](https://github.com/aws-samples/amazon-location-samples/blob/main/CONTRIBUTING.md) for more information.
+
+## License
+This library is licensed under the MIT-0 License. See the LICENSE file.
```
[the remark this answers is not in the document]
Comment (inariku) [reply maplibre-gl-js-vue-amplify-compare-maps/README.md:45]: Done

**Comment by inariku** [comment]
Thanks a lot for the pointers. I'm returning 'Done' comments on the ones I fixed. I have eliminated the margins on the map. The image is the app after the modification.
Screen Shot 2022-07-31 at 3 02 32

**Review by dreamorosi — CHANGES_REQUESTED** [review]
Please check that all files in the `.gitignore` are actually being ignored & not included in the commit. For example at the moment there's still one `amplify/backend/amplify-meta.json`.

This file cannot be in the repo as it has your account's details (i.e. ARNs, etc.).

**Comment by inariku** [comment]
Thank you. I had overlooked that.
I made sure there were no files included in the .gitignore.

File: maplibre-gl-js-vue-amplify-compare-maps/package.json [hunk 9]
```diff
@@ -0,0 +1,25 @@
+
```
Comment (dreamorosi) [inline maplibre-gl-js-vue-amplify-compare-maps/package.json:1]: Remove extra 2 lines on top

File: maplibre-gl-js-vue-amplify-compare-maps/package.json [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+{
+  "name": "maplibre-gl-js-vue-amplify-compare-maps",
+  "version": "0.1.0",
+  "private": true,
+  "scripts": {
+    "dev": "vite",
```
Comment (dreamorosi) [inline maplibre-gl-js-vue-amplify-compare-maps/package.json:8]: `start` (+ also update the README file)

File: maplibre-gl-js-vue-amplify-compare-maps/vite.config.js [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+
+// https://vitejs.dev/config/
+export default defineConfig({
+  plugins: [vue()],
+  resolve: {
+    alias: {
+      '@': fileURLToPath(new URL('./src', import.meta.url)),
+      './runtimeConfig': './runtimeConfig.browser',
+    },
+  },
```
Comment (dreamorosi) [inline maplibre-gl-js-vue-amplify-compare-maps/vite.config.js:14]: Please add the config [from here](https://github.com/aws-samples/amazon-location-samples/blob/main/maplibre-gl-js-vue-amplify/vite.config.js#L14-L17).

**Review by dreamorosi — CHANGES_REQUESTED** [review]
Hello, 

Thanks for making the changes, I have one last round of small changes (I promise) - see comments.

Also I still see the maps at two different heights, in multiple browsers:
Screenshot 2022-07-31 at 13 43 34


Screenshot 2022-07-31 at 13 43 12

**Comment by dreamorosi** [comment]
Okay, so I suggest we simplify things a bit:
- The focus of the sample should be to compare the maps side by side, so let's keep them always synced (& remove the toggle)
- Instead of providing an input field for the pitch, let's leverage the native components (added a `NavigationControl`) and/or the map behaviour (if you click your trackpad with two fingers and slide - it'll change the pitch)
- Right now you are running two styling components: Amplify UI and `element-plus`. Let's make Element Plus go away and use ours. Then with a few lines of CSS we can reproduce the same layout and remove extra padding + weird height behaviour (remove it also from dependencies & `package.json`). 

I managed to get it working 99%, there's only one thing that I'm not sure how to do which is how to make the select field selected by default with the right value when the page loads, I'm sure / I hope you can help with this.

See below:
- black circle around the pointer = single click (+ drag)
- green circle around pointer = double click (+ drag)
both modify the pitch:

https://user-images.githubusercontent.com/7353869/182033384-41743267-a625-4691-a9f2-1629977cf8cb.mp4


`src/main.js`
```js
import AmplifyVue from "@aws-amplify/ui-vue";
import "@aws-amplify/ui-vue/styles.css";
import "@maplibre/maplibre-gl-geocoder/dist/maplibre-gl-geocoder.css";
import { Amplify } from "aws-amplify";
import "maplibre-gl/dist/maplibre-gl.css";
import { createApp } from "vue";
import App from "./App.vue";
import awsconfig from "./aws-exports";

Amplify.configure(awsconfig);

const app = createApp(App);
app.config.productionTip = false;
app.use(AmplifyVue);
app.mount("#app");
```

`src/MapPane.vue`
```vue
<template>
  <div class="map-header">
    <div>
      <label>Map Style</label>
      <select v-model="renderedMap">
        <option
          v-for="map in availableMaps"
          :key="map.mapName"
          :label="map.style"
          :value="map.mapName"
        />
      </select>
    </div>
  </div>
  <div :id="id" class="map"></div>
</template>

<script>
import { createMap, createAmplifyGeocoder } from "maplibre-gl-js-amplify";
import { NavigationControl } from "maplibre-gl";
import { Geo } from "aws-amplify";
import { ref, watch, toRefs } from "vue";

export default {
  props: {
    id: String,
    availableMaps: Array,
    zoom: Number,
    center: Object,
    pitch: Number,
    bearing: Number,
    ActiveMap: String,
  },
  setup(props, context) {
    const renderedMap = ref(Geo.getDefaultMap());
    const map = ref(null);
    const { id, zoom, center, pitch, bearing, ActiveMap } = toRefs(props);

    const mapCreate = async () => {
      map.value = await createMap({
        container: id.value,
        zoom: zoom.value,
        center: center.value,
        pitch: pitch.value,
        bearing: bearing.value,
      });

      map.value.addControl(createAmplifyGeocoder());
      map.value.addControl(new NavigationControl({ visualizePitch: true }));

      map.value.on("mouseover", () => {
        context.emit("active-map-update", id.value);
      });

      map.value.on("move", () => {
        if (ActiveMap.value == id.value) {
          context.emit(
            "state-update",
            map.value.getZoom(),
            map.value.getCenter(),
            map.value.getBearing(),
            map.value.getPitch()
          );
        }
      });
    };

    mapCreate();

    watch(renderedMap, (renderedMap) => {
      map.value.setStyle(renderedMap);
    });

    watch(zoom, (zoom) => {
      if (ActiveMap.value !== id.value) {
        map.value.setZoom(zoom);
      }
    });

    watch(center, (center) => {
      if (ActiveMap.value !== id.value) {
        map.value.setCenter([center.lng, center.lat]);
      }
    });

    watch(bearing, (bearing) => {
      if (ActiveMap.value !== id.value) {
        map.value.setBearing(bearing);
      }
    });

    watch(pitch, (pitch) => {
      if (ActiveMap.value !== id.value) {
        map.value.setPitch(pitch);
      }
    });

    return {
      renderedMap,
    };
  },
};
</script>
```

`src/App.vue`
```vue
<template>
  <main>
    <header>
      <h1>Amazon Location × Amplify Geo</h1>
    </header>
    <section>
      <div>
        <Map
          id="left"
          :availableMaps="availableMaps"
          @state-update="updateState"
          @active-map-update="updateActiveMap"
          :zoom="zoom"
          :center="center"
          :pitch="pitch"
          :bearing="bearing"
          :ActiveMap="ActiveMap"
          :sync="sync"
        />
      </div>
      <div>
        <Map
          id="right"
          :availableMaps="availableMaps"
          @state-update="updateState"
          @active-map-update="updateActiveMap"
          :zoom="zoom"
          :center="center"
          :pitch="pitch"
          :bearing="bearing"
          :ActiveMap="ActiveMap"
        />
      </div>
    </section>
  </main>
</template>

<script>
import { Geo } from "aws-amplify";
import { ref } from "vue";
import Map from "./MapPane.vue";

export default {
  components: { Map },
  setup() {
    const availableMaps = ref(Geo.getAvailableMaps());
    const zoom = ref(14);
    const center = ref([139.7648, 35.6794]);
    const pitch = ref(30);
    const bearing = ref(0);
    const sync = ref(true);
    const ActiveMap = ref(null);

    function updateState(...args) {
      zoom.value = args[0];
      center.value = args[1];
      bearing.value = args[2];
      pitch.value = args[3];
    }

    function updateActiveMap(...args) {
      ActiveMap.value = args[0];
    }

    return {
      availableMaps,
      zoom,
      center,
      pitch,
      bearing,
      ActiveMap,
      updateState,
      updateActiveMap,
    };
  },
};
</script>

<style>
@import "@/assets/style.css";
</style>
```

`src/assets/style.css`
```css
body, body::after, body::before {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}

main {
  display: block;
  width: 100vw;
  height: 100vh;
  position: relative;
  overflow: hidden;
}

header {
  background: linear-gradient(225deg, #FE5151, #BD0916);
  width: 100%;
  height: var(--amplify-space-xxl);
  display: flex;
  align-items: center;
  justify-content: center;
}

header h1 {
  font-size: var(--amplify-font-sizes-xl);
  font-weight: var(--amplify-font-weights-light);
  color: var(--amplify-colors-white);
  margin: 0;
  padding: 0;
  text-align: center;
  line-height: var(--amplify-space-xl);
}

section {
  width: 100%;
  height: calc(100vh - var(--amplify-space-xxl));
  display: flex;
}

section > div {
  width: 50%;
  height: 100%;
}

section > div > .map-header {
  width: 100%;
  height: var(--amplify-space-xxl);
  display: flex;
  align-items: center;
  justify-content: center;
}

section > div > .map {
  width: 100%;
  height: calc(100% - var(--amplify-space-xxl));
}
```

**Comment by dreamorosi** [comment]
After the changes above are done - and @zachelliottwx gives us OK about the sample - I'll be happy to merge it 🎉

File: maplibre-gl-js-vue-amplify-compare-maps/package.json (hunk 10 above)
Comment (inariku) [reply maplibre-gl-js-vue-amplify-compare-maps/package.json]: Done

File: maplibre-gl-js-vue-amplify-compare-maps/package.json (hunk 9 above)
Comment (inariku) [reply maplibre-gl-js-vue-amplify-compare-maps/package.json]: Done

**Comment by inariku** [comment]
@dreamorosi Thank you. All the fixes are done!

@zachelliottwx Please check it !

File: maplibre-gl-js-vue-amplify-compare-maps/vite.config.js (hunk 11 above)
Comment (inariku) [reply maplibre-gl-js-vue-amplify-compare-maps/vite.config.js:13]: Done

**Comment by zachelliottwx** [comment]
There should be no problem with this. Looks good to me.

**Pull request merged**